### ISSUE TITLE
Clarify documentation

### ISSUE THREAD

**Issue by rrebollo** [odoo/upgrade-util#145]
https://github.com/odoo/upgrade-util/blob/0833a35e6858ff8b218631394f0e9ac05b258cb5/src/util/modules.py#L5-L6

Hi team. What do you mean by "a `base` script"? I mean I know script phases (pre|post|end). But I haven't heard of base scripts. Can U be more specific???

**Comment by aj-fuentes** [comment]
`base` in this context means the `base` _module_. The upgrade process in Odoo will always upgrade `base` _first_. Then the rest of the modules are upgraded, dependencies-first. Since `base` doesn't depend on any other module all module-related operations --like those in this file-- should be done in a script under `base` module. In that way, the upgrade of each module can proceed with the correct data in the DB.

In short: module level operations must be done in scripts under `base` module in order to prepare the rest of the upgrade.

**Comment by rrebollo** [comment]
@aj-fuentes thank U for the fast reponse but ... I don't get it ... Sorry. Since I don't work with core addons but custom ones: how can I use util.modules functions?

**Comment by rrebollo** [comment]
@aj-fuentes  your patience is gold. I have reviewed official docs, I respect your time (even if right now it looks like I don't :smile: ). In fact, I have already used upgrade utils in several migrations with success but module functions are particularly evasive. So putting all together I think U are suggesting emulate a directory tree like (attached image) and pointing to it with upgrade-path parameter:
![image](https://github.com/user-attachments/assets/55905964-39b1-43c1-8d40-f1d99d89516c)
Am I right? I don't think it is obvious according to docs. When U told me about "base" module I was thinking in Odoo base addon. I should be explicit about I'm using upgrade utils not only for major version upgrades but minor ones too: things like x.0.1.0.0 to x.1.1.0.0. Is this a valid scenario for upgrade utils?? And lastly is "base" script like a convention? because it doesn't even look like a python module (no __init__.py)
![image](https://github.com/user-attachments/assets/af97a38d-dd30-4321-92fb-099324f07f0f)
Thank U so much for all the guidance. Sorry to be so annoying.

**Comment by aj-fuentes** [comment]
> when U told me about "base" module I was thinking in Odoo base addon. I should be explicit about I'm using upgrade utils not only for major version upgrades but minor ones too: things like x.0.1.0.0 to x.1.1.0.0.

I was referring to the `base` module of Odoo. Now, it's true that there is a flaw in the flow: you cannot trigger a `base` upgrade script without `base` being upgraded, which is not something you can do post upgrade. I guess we will need to improve the documentations in that regard.

> Is this a valid scenario for upgrade utils?? 

Yes.

> And lastly is "base" script like a convention? because it doesn't even look like a python module (no init.py)

The directory structure is a way to organize scripts. It doesn't matter _where_ they are. They are all taken into account and run by the upgrade system in the code of Odoo. You can specify as many `upgrade-path`s as you want. As long as all follow the `<module>/<version>/<prefix><whatever>.py` structure, they will all be executed. Thus there is no need for an `__init__.py` anywhere.

Concrete example, imagine this structure:
```
~/my-upgrade
  +--my_module
     +--18.0.0.2
        +--pre-migrate.py
        +--post-migrate.py
  +--my_module2
     +--17.0.1.2
        +--end-migrate.py
~/other-scripts
  +--my_module3
     +--16.0.1.0
        +--pre-migrate.py
```
You upgraded your DB with Odoo's upgrade platform from 14 to 18. Then you want to upgrade your custom modules `my_module,my_module2,my_module3`. You should then run Odoo 18 with:
```sh
./odoo-bin -u all --addons-path=... --upgrade-path=~/my-upgrade,~/other-scripts --stop
```
Then you should see _your_ modules being upgraded following the versions in the scripts folder. 

Here is the flaw: you cannot trigger a `base` script even if you put a script in the right dir because `base` is already in the final version. In this sense I agree with you that scripts for module-operations are tricky to use. In principle, you _can_ use those utils anywhere. The details I referred to [here](https://github.com/odoo/upgrade-util/issues/145#issuecomment-2388434218) are related to _why_ the expected way to use them is in `base`. You can even use them by starting Odoo in shell mode and calling the utils directly.
```py
>>> from odoo.upgrade import util
>>> util.remove_module(self.env.cr, "my_module")
```
The alternative is to use it in `pre-` of the first of your custom modules that's being upgraded. This is far from ideal, I agree.

I took note of this flaw and we'll see if there is a way we can adapt the core mechanism to run scripts before any other module is upgraded. Not necessarily as part of a `base` module upgrade. In the meantime the documentation could be adapted to reflect better these shortcomings.

**Comment by aj-fuentes** [comment]
I realize that there may be also a misunderstanding about what `base` means. [Here](https://github.com/odoo/odoo/tree/18.0/odoo/addons/base) is the _actual_ `base` module.

On a final note. We _do_ want you (the community) to use the utils for the upgrade of your custom modules. Thus we are always happy --time allowing-- to clarify any technical issue. The documentation would be gradually improved. Thanks for reporting any such issue.

**Comment by lem8r** [comment]
> That would be done in a script located at `path-to-your-src/base/<target-version>.1.3/pre-my-upgrade.py`

Module level operations are working only in post- scripts in my case.
In pre-migtaions there is an error `AttributeError: module 'odoo.modules' has no attribute 'remove_module'`

Also after we got migrated db it's base module already have `<target-version>.1.3` so it's version has to be changed back in order to trigger our custom `base` scripts.

**Comment by lem8r** [comment]
> ```shell
> ./odoo-bin -u all --addons-path=... --upgrade-path=path-to-utils,path-to-your-src --stop
> ```

Passing `UPG_AUTOINSTALL='all'` is also important for module level migrations
> ```shell
> $ UPG_AUTOINSTALL='all' ./odoo-bin -u all --addons-path=... --upgrade-path=path-to-utils,path-to-your-src --stop
> ```

**Comment by aj-fuentes** [comment]
Related: https://github.com/odoo/odoo/pull/202014
